### PR TITLE
Add proper type annotations for `array like` + usage examples

### PR DESCRIPTION
**Pull request by idanmiara**
Closes https://github.com/shapely/shapely/issues/1683
This PR is a proposed solution for adding proper type annotations for `array like` and also replace the current documentation statements like `Geometry or array_like` with 2 usage examples.

### CODE REVIEW

**Comment by idanmiara** [comment]
@jorisvandenbossche - I know this is a huge PR that would need to be broken down to pieces... 😵‍💫 
But could you give me heads up what's your general take on these ideas ?

**Comment by idanmiara** [comment]
@jorisvandenbossche any comments on this PR please?

**Comment by brendan-ward** [comment]
Thank you for your work on this @idanmiara and apologies for the slow review of PRs.  Your effort toward improving Shapely is appreciated!

Adding type annotations is a big addition that needs to be weighed carefully and needs buy-in from several of the Shapely maintainers, because this adds to the overall maintenance burden of the project as well as potentially making it harder for new contributors, especially for those of us that don't use them on a regular basis.  This is big enough that it probably warrants taking it up first as a discussion in order to get buy-in, rather than via PR review here.

Even without a close review of the changes here, I think this needs to be split up a bit differently into smaller, more focused PRs.  Specifically, documentation changes that are unrelated to type annotations should be addressed in a separate PR; those should theoretically be easier to review / merge.  I'd recommend focusing on the documentation changes first; I'm hoping those are easy to copy / paste out of here.  Apologies that this involves extra work for you.

In general, small & focused PRs are more manageable to review given limited resources; big & complex PRs are very challenging to review and often take more time to review than folks have available - which means that they don't get reviewed at all for long periods of time.  I can't speak for other Shapely contributors here, but my experience has been that if I can't review a PR within a narrow window of time that I have available, I often can't get to it at all in a reasonable amount of time - which is surely frustrating to folks that are actively trying to make the project better like yourself.

**Comment by jorisvandenbossche** [comment]
As @brendan-ward mentioned, I think it would be good to first have some discussion about type annotations in an issue, to see how the other maintainers think about it (I personally also have reservations, but won't object). 
Previous discussion about this: https://github.com/shapely/shapely/issues/721

And also if we are adding type annotation, I would highly recommend starting with annotating a small part so that it is easier to review (I expect that the first PRs will generate quite some back and forth about the exact form of the type annotations that we want, and once that is somewhat agreed upon, subsequent PRs expanding the coverage should go much smoother).

**Comment by jorisvandenbossche** [comment]
FWIW I personally also prefer that we keep annotation syntax just for annotations, and not start using it in docstrings (IMO a docstring with semi-fluent text like "Geometry or array-like of geometries" reads much easier and is more understandable than "MaybeGeometryArrayNLike")

**Comment by idanmiara** [comment]
> As @brendan-ward mentioned, I think it would be good to first have some discussion about type annotations in an issue, to see how the other maintainers think about it (I personally also have reservations, but won't object). Previous discussion about this: #721
> 
> And also if we are adding type annotation, I would highly recommend starting with annotating a small part so that it is easier to review (I expect that the first PRs will generate quite some back and forth about the exact form of the type annotations that we want, and once that is somewhat agreed upon, subsequent PRs expanding the coverage should go much smoother).

@jorisvandenbossche thanks for this pointer!
Would you want to reopen #1721?
I don't mind making disecting this PR - would you like to propose which section do you think would be the best to discuss first? 
This PR is a combination of two efforts :annotations + documentation consolidation.
I'm not sure what would be a good way to split.

**Comment by jorisvandenbossche** [comment]
> This PR is a combination of two efforts :annotations + documentation consolidation.
> I'm not sure what would be a good way to split.

I think annotations and documentation changes are completely orthogonal? (assuming we don't use annotation-like syntax in the docs) 
So that is an obvious split I would say. 

> would you like to propose which section do you think would be the best to discuss first?

I would start with a single file, and base.py might be the easiest to start with (and also the one with potentially most value)

**Comment by idanmiara** [comment]
> I think annotations and documentation changes are completely orthogonal? (assuming we don't use annotation-like syntax in the docs) So that is an obvious split I would say.

Agree, but there would be many conflicts to solve if we handle both issues concurrently (editing sections close to each other).
Thus I'm thinking maybe to do the doc consolidation effort first while we discuss the annotations on a small scale.

How would you advise to split the consolidation PR? 
I'm thinking, once https://github.com/shapely/shapely/pull/1673 is merged I'd apply the same logic for all the functions and split it into a few PRs - I'm was thinking to split it maybe like this:
* all the small fixes in a single PR
* some bigger changes like `snap` each in a separate PR
* then maybe a PR per module

@jorisvandenbossche - What do you think?

> > would you like to propose which section do you think would be the best to discuss first?
> 
> I would start with a single file, and base.py might be the easiest to start with (and also the one with potentially most value)

Sure, no problem.

**Comment by sgillies** [comment]
@idanmiara I don't understand why you take on very big new efforts in this project instead of working on the many smaller issues? We have 100+ open issues in this project.

**Comment by idanmiara** [comment]
> @idanmiara I don't understand why you take on very big new efforts in this project instead of working on the many smaller issues? We have 100+ open issues in this project.

Hi @sgillies,
Thanks for asking 😄 here is a my non-exhaustive list of reasons that came to my mind:

> ... instead of working on the many smaller issues? We have 100+ open issues in this project.

Many times I take smaller existing issues and contribute for the following reasons:
1. Reviewing to solve existing PRs to help other contributors: https://github.com/shapely/shapely/pull/1231 https://github.com/shapely/shapely/pull/1448 https://github.com/shapely/shapely/pull/1685. [I believe that we need more contributors]
(https://github.com/shapely/shapely/issues/1702) and I believe that reviewing PRs help bring in new contributors.
2. Issues that I think that solving them would make it easier for others to contribute because things would be more 
organized, like https://github.com/shapely/shapely/pull/1681
3. Issues that fail tests and [make it seems like there is a bug when there isn't](https://github.com/shapely/shapely/pull/1655#issuecomment-1370008684), like https://github.com/shapely/shapely/pull/1715 
3. For the sake on learning, helping others or giving back to the community, like https://github.com/shapely/shapely/pull/1720 - I couldn't care less about supporting input`decimal`s 😆 , but I understood it does matter to for other people so to mark it as fix in `2.0.1`.
4. Smaller Issues that effect others even if might not effect me personally, like https://github.com/shapely/shapely/pull/1721 https://github.com/shapely/shapely/pull/1695, https://github.com/shapely/shapely/pull/1708
5. Issues that effect my workflow significantly, bug me personally or slow me down when working on other issues that do matter, like https://github.com/shapely/shapely/pull/1703,  https://github.com/shapely/shapely/pull/1694,  https://github.com/shapely/shapely/pull/1682
7. Issues that effect performance of critical parts of my code, like https://github.com/shapely/shapely/pull/1676 https://github.com/shapely/shapely/pull/1722

> I don't understand why you take on very big new efforts...

1. As for this specific issue - when [I want to learn a new library](https://github.com/rasterio/rasterio/pull/2200) I often read the documentation or go over the code. If the documentation is fragmented or hard to understand I am trying to fix parts that I was able to figure out so it would be easier for others and also for myself for future reference, this helps me understand better how things work. That was also the case with https://github.com/shapely/shapely/pull/1673. Then I saw that the fragmented documentation is all over because of the merge of `pygeos` into `shapely` so I wanted to consolidate the documentation so it would be easier for me to [find the right documentation when I need it](https://github.com/shapely/shapely/issues/1668#issuecomment-1351275073). Sometimes it worth for me to do it in a large scale all at once if I believe I would need to read and use large parts of the library - so I went on a quest to consolidate the documentation as I also believe this would help others to use the shapely better and more easily.
2. But I'm guessing that you were referring to the type annotations major effort:[ for me they play a MAJOR part in making the documentation readable](https://github.com/rasterio/rasterio/issues/2322): Type annotations make it easier to understand what are the appropriate inputs and outputs, both programmers and for IDEs and automated tools. For instance, I am working on performance boost for some internal code, thus I need to understand better what are all the options for proper inputs and outputs for the different functions (https://github.com/shapely/shapely/issues/1683). This also help to find bugs in internal code as the IDE and automate tools points point out if the inputs/outputs for a function are wrong.

So to summarize - for me - this issue is one of the most important one to solve (right after solving regressions), and worths my time spending on solving it. 
I'm positive that this effort would make it MUCH easier other users ([and especially new users and contributors](https://github.com/shapely/shapely/discussions/1316)) to make the best out of shapely.
I'm certainly onboard helping with [your concerns about them](https://github.com/shapely/shapely/issues/721#issuecomment-1407212866).

**Comment by HaraldTR** [comment]
This is fantastic work imo.
